### PR TITLE
CDAP-15516: Support Record type for BigQuery Source and Sink plugins

### PR DESCRIPTION
**Pull request by aielchiieva**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -154,6 +154,20 @@
       <groupId>com.google.cloud.bigdataoss</groupId>
       <artifactId>bigquery-connector</artifactId>
       <version>${bigquery.connector.hadoop2.version}</version>
+      <exclusions>
```
Comment (albertshau) [inline pom.xml:160]: can you add a comment about why these need to be excluded

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java [hunk 2]
```diff
@@ -318,18 +318,38 @@ private void validateSchema(BigQuery bigQuery, String tableName,
...
+      List<Schema.Field> schemaFields;
+      if (Schema.Type.ARRAY == field.getSchema().getType()) {
+        schemaFields = Objects.requireNonNull(field.getSchema().getComponentSchema()).getFields();
+      } else {
+        schemaFields = field.getSchema().isNullable()
+          ? field.getSchema().getNonNullable().getFields()
+          : field.getSchema().getFields();
+      }
+      fieldSchema.setFields(Objects.requireNonNull(schemaFields).stream()
+              .map(this::generateTableFieldSchema)
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java]: nit: indentation

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java [hunk 3]
```diff
@@ -167,11 +171,18 @@ private static void writeArray(JsonWriter writer, String name, Object value, Sch
...
-    if (!(value instanceof Collection) && !value.getClass().isArray()) {
-      throw new IllegalArgumentException("The value should be of type collection or array. Got: " + value.getClass());
+    Collection collection;
+    if (value instanceof Collection) {
+      collection = (Collection) value;
+    } else if (value instanceof Object[]) {
+      collection = Arrays.asList((Object[]) value);
+    } else {
+      throw new IllegalArgumentException(String.format(
+        "The value is of type '%s', expected Collection or array.", value.getClass().getSimpleName()));
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: This is an existing issue, but the message should include the name of the field.

A value for field '%s' is of type '%s' when it is expected to be a Collection or array.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java [hunk 4]
```diff
@@ -180,28 +191,49 @@ private static void writeArray(JsonWriter writer, String name, Object value, Sch
...
   }
 
+  private static void writeRecord(JsonWriter writer, String name, Object value, Schema fieldSchema) throws IOException {
+    if (value == null) {
+      writer.name(name);
+      writer.nullValue();
+      return;
+    }
+
+    if (!(value instanceof  StructuredRecord)) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: extra space after instanceof

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java [hunk 5]
```diff
@@ -180,28 +191,49 @@ private static void writeArray(JsonWriter writer, String name, Object value, Sch
...
-        write(writer, name, true, Array.get(value, i), componentSchema);
+        writer.endObject();
+      } else {
+        write(writer, name, true, element, componentSchema);
       }
     }
     writer.endArray();
   }
 
+  private static void writeRecord(JsonWriter writer, String name, Object value, Schema fieldSchema) throws IOException {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: annotate value as nullable

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java [hunk 6]
```diff
@@ -180,28 +191,49 @@ private static void writeArray(JsonWriter writer, String name, Object value, Sch
...
+                                                           "which is not allowed by BigQuery.", name));
       }
-    } else {
-      for (int i = 0; i < Array.getLength(value); i++) {
-        // big query does not allow null values in array items
-        if (Array.get(value, i) == null) {
-          throw new IllegalArgumentException(String.format("Field '%s' contains null values in its array, " +
-                                                             "which is not allowed by BigQuery.", name));
+      if (element instanceof StructuredRecord) {
+        writer.beginObject();
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: this block of code and the last part of the writeRecord() method are the same. Let's create a new method that is called in both places.

**Review by albertshau — COMMENTED** [review]
minor comments.

Can you also create an integration test in https://github.com/cdapio/cdap-integration-tests? (After this PR is merged)

**Comment by albertshau** [comment]
Also, can we merge this to the release/0.12 branch instead of develop

File: pom.xml (hunk 1 above)
Comment (aielchiieva) [reply pom.xml:160]: Done.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java (hunk 2 above)
Comment (aielchiieva) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java]: Done.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java (hunk 3 above)
Comment (aielchiieva) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: Done.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java (hunk 4 above)
Comment (aielchiieva) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: Removed.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java (hunk 5 above)
Comment (aielchiieva) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: Done.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java (hunk 6 above)
Comment (aielchiieva) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryRecordToJson.java]: Done.

**Comment by aielchiieva** [comment]
@albertshau addressed code review comments, please take a look.

File: pom.xml [hunk 7]
```diff
@@ -154,6 +154,21 @@
...
+        <!-- Dependencies are excluded because they are missing in central Maven repository -->
```
Comment (albertshau) [inline pom.xml:161]: weird, so it doesn't compile unless you exclude these?

File: pom.xml [hunk 8]
```diff
@@ -403,6 +418,22 @@
...
+    <!-- Dependencies are required after updating bigquery.connector version from 0.13.1-hadoop2 to hadoop2-0.13.17 -->
+    <dependency>
+      <groupId>commons-codec</groupId>
+      <artifactId>commons-codec</artifactId>
+      <version>1.4</version>
+    </dependency>
+    <dependency>
+      <groupId>com.google.flogger</groupId>
+      <artifactId>flogger-system-backend</artifactId>
+      <version>0.3.1</version>
```
Comment (albertshau) [inline pom.xml]: can you add properties for these versions

**Review by albertshau — COMMENTED** [review]
minor comment about use properties for versions in the pom, otherwise lgtm. Fix, squash, and we'll merge.

File: pom.xml (hunk 7 above)
Comment (aielchiieva) [reply pom.xml:161]: Correct.

File: pom.xml (hunk 8 above)
Comment (aielchiieva) [reply pom.xml]: Done.

**Comment by aielchiieva** [comment]
@albertshau made final changes and squashed the commits.

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**